### ISSUE TITLE
Add integration testing for Oracle, for official DBMS support

### ISSUE THREAD

**Issue by Ste1io** [CollaboratingPlatypus/PetaPoco#699]
PetaPoco has "unofficially" supported Oracle for quite some time now... I'd like to implement the integration tests for Oracle so that can change finally, and we can claim official support without a big *"but"* after. This will also allow us to confidently investigate, address, and close multiple issues that are unclosed, and can most likely be readily fixed provided tests are possible.

Happy to handle this myself alongside the remaining test refactoring.

Closes #691
Addresses #626
Addresses #613

**Comment by Curlack** [comment]
Busy setting up everything. Have used Oracle for a while couple of years
back. Might be ready to lend a hand soon 😁

**Comment by Ste1io** [comment]
Let me know how setup goes with your local environment, @Curlack, and if you encounter anything you think should be added to the updated wiki page. 

Given your background in Oracle, I don't mind passing this one over to you if you want, or collaborating together on it, whichever is best for you.

**Comment by Curlack** [comment]
Setup was smooth. After following all the instructions. The code compiled
and all tests passed 😎.
Collaborating at first at least. Need to find my bearings

**Comment by Curlack** [comment]
I'm gonna need the docker setup for oracle. I can see a comment for oracle pointing to "sath89/oracle-12c:latest" image, but to be honest, I have no clue how all the docker stuff works. Will figure it out as I go. Not sure if we'll start with 12c and go up to 23c or for all intents and purposes 12c will always be enough.

PS: Also wanted to ask if this will be our primary medium of communication and (since we're 6 hours apart) what times would suite you best?

**Comment by Ste1io** [comment]
For the docker images, most of it's been trial and error and checking docs... the sathu89 image (that commented service predates my recent PR) is no longer available when I checked (apparently due to a dmca from Oracle ~2019). afaik, 19c is current LTS, so that would be my suggestion. Including 12c for backwards compatibility with the older Oracle provider (the orphan in the root directory, I suppose) may be necessary.

Matrix testing hasn't really been brought up yet, but you'll notice I added comments about firebird also as the current image used for it is a couple major versions behind, like 12c.

Just noticed I neglected to update the .vscode extensions with it, but if you install the the [Docker extension](https://marketplace.visualstudio.com/items?itemName=ms-azuretools.vscode-docker) for VSCode, testing images is fairly straightforward simply by right-clicking on the compose.yml after making your changes and selecting the services you want to fire up.

**Comment by Curlack** [comment]
So far I've been hitting brick walls. I'm trying to build a docker image using [Docker Images for Oracle repository](https://github.com/oracle/docker-images). Should be simple only need to execute buildContainerImage.sh file, but keep getting "not found" error in terminal. My VSCode session is connected to WSL and Desktop Docker is running. Not sure what I'm missing.
Want to build the "Oracle Database Free" (23c) image with 19c compatibility. Any pointers would be welcome.

**Comment by Curlack** [comment]
Success at last! Used installation described on the [Oracle Container Registry](https://container-registry.oracle.com/). Opted for version 23c (free). Also installed [Oracle SQLDeveloper](https://www.oracle.com/database/sqldeveloper/) ide (less "consoley" hehehe). Turns out compatibility has nothing to do with database features, only disk format. Also I just needed to read the same thing enough times for it to "click" :)

The command I ended up using is:
```
docker run --name oracle-free-v23c -d -p 1522:1521 -p 5700:5500 -e ORACLE_PWD=petapoco -e ORACLE_CHARACTERSET=AL32UTF8 container-registry.oracle.com/database/free:latest
```
One thing I'm not sure of, is how the `docker-compose.yml` file should be changed.

Now that I'm able to connect I can maybe begin with small scaffold of oracle test group and work my way up as I go.

**Comment by Ste1io** [comment]
Nicely done @Curlack. Translating from cmd line to yml is typically fairly straightforward (-e as environment variables, -p being ports). Something like this should work, added under the services...untested:

```yml
oracle-free-v23c:
  image: container-registry.oracle.com/database/free:latest
  container_name: oracle-free-v23c
  environment:
    - ORACLE_PWD=petapoco
    - ORACLE_CHARACTERSET=AL32UTF8
  ports:
    - "1522:1521"
    - "5700:5500"
```

**Comment by Ste1io** [comment]
> Now that I'm able to connect I can maybe begin with small scaffold of oracle test group and work my way up as I go.

I'm waiting on @pleb to review my changes to the tests, but you should be safe forking off of #701 and building up the oracle tests from there. The SQLite tests, and the accompanying test providers for SQLite provide a fairly straightforward bare-bones multi-driver implementation. As with the other tests, instantiate a class inheriting from one of the abstract classes to activate those tests; override or skip any as needed, or add additional test methods if specific to oracle.

If you want me to help scaffold an initial oracle test class just lmk.

**Comment by Curlack** [comment]
Quick update from my side. Btw, thanks for the branch, good idea.
`docker-compose.yml` updated and working (thanks for the tip). Beginning to understand docker a little better now.
`app.config` and `appsettings.json` file changes in place and oracle database connection successful. Using [Oracle.ManagedDataAccess 21.12.0](https://www.nuget.org/packages/Oracle.ManagedDataAccess) nuget package.
Created `OracleTestProvider.cs`, generated a setup script and "et voila!" I was ready to start converting build scripts to their oracle equivalent....until I got to the `Stored Procedures` :-(.

```sql
CREATE PROCEDURE dbo.SelectPeople
AS
BEGIN
	SET NOCOUNT ON;
	SELECT * FROM [People]
END
```
This simple procedure returns a dataset as if it's a table. Only difference is we execute it instead of selecting from it. After doing some reading and even consulting Bing AI, they all mention an output parameter of type `Ref Cursor`, but the caller has to explicitly bind the parameter.

As of v12c and above they boast to be able to return query results to the caller implicitly via [DBMS_SQL.RETURN_RESULT](https://docs.oracle.com/database/121/LNPLS/dynamic.htm#GUID-8CFE5AA9-0BBD-417B-A7EE-6F05AF49001B) and [DBMS_SQL.GET_NEXT_RESULT](https://docs.oracle.com/database/121/LNPLS/dynamic.htm#GUID-1D6A4302-CBA0-47A6-B4A7-80B089DF4E61).

I suppose if it's possible  to do by any means, it's fine as long as the end result prevents the caller from modifying the underlying table, which is why these kind of procs exist in the first place imo. Just having bit of a tough time trying to get around this one.

Any ideas or alternatives guys?

**Comment by Ste1io** [comment]
Nicely done, @Curlack. 

My initial inclination would be to implement an Oracle-specific override of

https://github.com/CollaboratingPlatypus/PetaPoco/blob/53e928c6dc44819aa834683abcd108029f960b77/PetaPoco/Database.cs#L990-L994

`Ref Cursor` would need passed in, and of course this would also require overrides of the Fetch/Query helpers as well, Then inside, push the cursor param containing the output from the stored proc into our command...

A naïve approach being something to the affect of

```cs
using (var cmd = CreateCommand(_sharedConnection, commandType, sql, args))
{
    if (_provider is Providers.OracleDatabaseProvider && commandType == CommandType.StoredProcedure)
    {
        var refCursorParam = /* create parameter here */;
        cmd.Parameters.Add(refCursorParam);
    }
    //...
```

So given the stored procedure in question,

```sql
CREATE OR REPLACE PROCEDURE SelectPeople (p_cursor OUT SYS_REFCURSOR)
AS -- or IS, not sure off hand
BEGIN
    OPEN p_cursor FOR
    SELECT * FROM People;
END;
```

We'd end up with something akin to

```cs
[Fact]
public virtual void QueryProc_NoParam_ShouldReturnAll()
{
    var results = DB.QueryProc<Person>("SelectPeople", "p_cursor").ToArray();
    results.Length.ShouldBe(6);
}
```

**Comment by Ste1io** [comment]
> `docker-compose.yml` updated and working (thanks for the tip). Beginning to understand docker a little better now.
> `app.config` and `appsettings.json` file changes in place and oracle database connection successful. Using [Oracle.ManagedDataAccess 21.12.0](https://www.nuget.org/packages/Oracle.ManagedDataAccess) nuget package.

When you get a chance, if you want to open a PR against the upstream's oracle feature branch with these changes (excluding the unfinished changes being discussed, ofc), I'll pull them in so we're both working in the same test environment.

**Comment by Curlack** [comment]
What I was aiming for was to also have working build scripts in place
before the PR. So the database is not only setup, but also in starting
state. Off course a lot of the tests might/will fail, but at least step 1
in typical TDD done :-)

If you're happy with ref cursor, I'll finish the rest of the build script
then submit a PR.

Thanks so much.

**Comment by Ste1io** [comment]
> If you're happy with ref cursor, I'll finish the rest of the build script then submit a PR.

Using `Ref Cursor` seems to me the least invasive option and most viable direction given the scope of this branch, and should be implementable by extending the current code without introducing any major breaking changes to the API.

I do think it would be prudent to leave the door open to the newer approach, however, I see that more as a future extension of the library's Oracle support than a requirement for official support.

You have more experience working with Oracle than me, also, so if you feel strongly either way, I'm keen to hear. I'd like to hear @pleb and @asherber's thoughts as well.

**Comment by Curlack** [comment]
I agree. On further investigation, it seems the C# implementation is the
same either way. So I'm gonna go with that. Code base will, inadvertently,
cater for the new way as well....developers dream!

**Comment by pleb** [comment]
Love your work @Ste1io and @Curlack.

Don't forget to update the readme for oracle support and add yourselves to the `PetaPoco is maintained and extended by` line :)

**Comment by Ste1io** [comment]
> Don't forget to update the readme for oracle support

There's a few other places throughout the docs as well I'm in the process of updating already, Will be sure to do that. @Curlack hammered this out fast, very excited with the progress.

**Comment by Curlack** [comment]
Always a pleasure. Glad I'm able to contribute to such an awesome codebase. 😎

**Comment by Curlack** [comment]
Quick update: I've added the integration tests for Oracle and ran the tests. Couple of them passed, but the majority failed. But we expected as much...

Looking into the first failure: `Delete_GivenPoco_ShouldDeletePoco`, very first line i.e. the Insert(person), I had `Oracle.ManagedDataAccess.Client.OracleException : ORA-01465: invalid hex number`. This is most likely due to `RAW(16)` data type used for storing `Guid`.

As I'm debugging, I saw the following insert statement:

```#sql
INSERT INTO "PEOPLE" ("ID","FULLNAME","AGE","HEIGHT","DOB") VALUES (:0,:1,:2,:3,:4)
```

I found that the default implementation of `EscapeSqlIdentifier` in the `PetaPoco.Providers.OracleDatabaseProvider` is confusing the issue.
An identifier is either **delimited** (surrounded by double quotes), which makes it case sensitive, or **ordinary** which has limits (must start with character, can only contain letters, digits and underscore), but is case insensitive.

```c#
public override string EscapeSqlIdentifier(string sqlIdentifier) => $"\"{sqlIdentifier.ToUpperInvariant()}\"";
```

By default oracle objects will be created in uppercase unless surrounded by double quotes. The the user is pretty much in complete control of naming tables, columns, etc.

Shortcut would be to just return the identifier that was passed in, but I can also use Regex to determine whether quotes are required and include them if not already included.

It seems the intention was to automagically give the user the correct syntax, so obviously I'm leaning towards the latter.
The question is what do you guys prefer @Ste1io, @pleb, @asherber ?

PS: Technically this me getting distracted, so just wanted to mention this, put a FIXME or TODO and continue.

**Comment by Ste1io** [comment]
Great timing, @Curlack. Was just getting on to check if you had started on that or wanted me to get rolling on that.

> Shortcut would be to just return the identifier that was passed in, but I can also use Regex to determine whether quotes are required and include them if not already included.

For now, I'm leaning towards adding a TODO, and changing the `EscapeSqlIdentifier` for Oracle to essentially be a no-op, as you mentioned. We can sort out how best to handle that between now and finishing the remaining tests. Personally, I'm not that keen on using regex for this; even if compiled; that should be a fairly simple check?

On that same note, I don't believe there's any error checking going on for multiple escaping, or if that's even something that's an issue to begin with. Figured I'd mention it nonetheless, since Oracle's use of quotes it seems could become problematic if double-escaped, though I could be wrong. e.g., (using a legal but discouraged table name): `SELECT * FROM ""Foo Table""`.

**Comment by Ste1io** [comment]
btw @Curlack, so you're aware and to prevent any unnecessary work... I'm currently wrapping up some general cleanup across the tests in a separate private branch, proofing my doc comments from earlier, and also addressing my inline todo's (those that don't involve breaking changes).

**Comment by Curlack** [comment]
So the "invalid hex number" issue was due to the hypens in the value.
1. Either `OracleDatabaseProvider` needs to handle `Guid` a little different for `RAW(16)` data type i.e. `Id.ToString("N")` instead of `Id.ToString()` **OR** column type should be `VARCHAR2(36)` i.e. we're not going to support the recommended `UNIQUEIDENTIFIER` data type. Telling the OracleDatabaseProvider it `HasNativeGuidSupport` doesn't work.
2. `DATETIME` is also finicky. `OracleDatabaseProvider` needs to use `TO_TIMSTAMP` or `TO_TIMESTAMP_TZ` functions and be sure to at least abbreviate the month in the format string to not confuse day for month in less than 13 cases lol.

The following sql successfully inserted the record.
```sql
INSERT
    INTO PEOPLE
        (ID,
        FULLNAME,
        AGE,
        HEIGHT,
        DOB)
    VALUES
        ('9b3e6f88d0eb4fa387fad7a33b09d2d9',
        'Peta',
        18,
        180,
        TO_TIMESTAMP_TZ('1945-DEC-01 05:09:04 -8:00', 'yyyy-mon-dd hh24:mi:ss tzh:tzm'))
```

I'll make the "noop" change and push the tests as-is. Then we divide and conquer, unless you also want me to change the db scripts based on the route we decide to take for #1 above?

**Comment by Ste1io** [comment]
Divide and conquer works for me. If you don't see any issues with how the db scripts are currently, no need change them until we need to. I'd err on the side of "best practice" when it gets to db scripts in general.

`VARCHAR2(36)` sounds like the right approach to me if we're dealing with the GUIDs as strings... `9b3e6f88d0eb4fa387fad7a33b09d2d9` appears to be the hexadecimal conversion of the string. What about working with GUIDs as 16-byte arrays? That's commonly used for GUID storage, and might be what oracle is wanting in this case?

Re timestamps, sounds like `DateTimeOffset` might be appropriate here, beyond the script usage... Need to double check how the other db's are handling datetime fields again as a refresher.

```cs
dateTimeOffset.ToString("yyyy-MMM-dd HH:mm:ss zzz", CultureInfo.InvariantCulture).ToUpper();
```

**Comment by Curlack** [comment]
Ok cool. I changed the db scripts to use `VARCHAR2(36)`, I agree with you, keep it simple. We won't have any issues with primary keys and referential integrity, but it will be case sensitive. Small price to pay considering it's all handled by the system.

Good news: over 60% of the tests pass.
Bad news: the ones that are failing, are tough.

**Comment by Ste1io** [comment]
> Good news: over 60% of the tests pass.
> Bad news: the ones that are failing, are tough.

Easy. New game plan:

![zsqg5x9d8pr11.jpg](https://github.com/CollaboratingPlatypus/PetaPoco/assets/37424493/1faf7e85-fb24-42c7-8696-56307b17444c)

**Comment by Ste1io** [comment]
I'm just stepping out for dinner @Curlack; will sync up locally when I get back and run the tests so I can see where things stand. If there's any particular collection or test you want me to focus on lmk. I'll be free for the rest of the evening into late night after eating so hopefully we can make a dent in things.

**Comment by Ste1io** [comment]
Getting a later start than I intended, but I'm here and synced up. 70 of 302 failing tests... Not bad for a first go at it tbh.  A couple first impressions after a very brief cursory glance over the failed tests...

Two patterns are standing out to me: storedproc tests are all failing with only a couple of exceptions, and nearly all dynamic type tests are failing. Not all dynamic type tests are failing, so I doubt dynamic type is the cause... But one commonality with those is statically typed table names in the SQL strings. This is something I've been fixing in my tests cleanup branch (still local at the moment). But regardless, they all use the standard pascal casing, and based on Oracle docs, reiterated by you earlier in this thread, a fair number of those tests are probably using uppercase table names.

I'm going to follow my gut and explicitly quote table names in the script and see if it has any effects on the test run, and go from there. Also going to write some tests for the provider class escaping methods since there are none currently that I recall, and that needs ruled out as a possible factor at play here.

**Comment by Curlack** [comment]
For the case sensitivity issue. I had the scripts like that (quoted
identifiers) before, not realizing the issue with `EscapeSqlIdentifier`.
In hindsight, only uppercasing would've been the issue then.
In order to cater for all these flavors of database objects, it's gonna get
messy very soon.

From my experience using Oracle, Ordinary Identifiers are preferred i.e. no
quotes. This makes writing any script less prone to case mismatches.

To give you a picture of the prevurbial can of worms, see the difference
between `select Column from Table` vs `select * from Table` vs `select
"Column" from "Table"`.

The other sad part of the story is that it's not ideal to tell someone they
have to change their whole database and all the applications talking to it,
if they wanna use PP :( lol.

**Comment by Curlack** [comment]
For the dynamic objects, we have a couple of options:
1. Make the underlying dictionary, of the ExpandoObject, case insensitive.
2. Produce the correct casing on creation by convention.
3. Query the expected casing from the result based on db implementation
flavor.

I've had huge success using Expando instead of ExpandoObject in one of my
projects. See
https://weblog.west-wind.com/posts/2012/feb/08/creating-a-dynamic-extensible-c-expando-object
.
Not sure if this can help us here though.

Gonna do some research on this later.

Wanna tackle the paging queries needing a table alias first. Thinking to
use string templating algorithm utility method in either new class or
ParameterHelper utility.

**Comment by Curlack** [comment]
Currently the `OracleDatabaseProvider` is reusing `BuildPageQuery` from `SqlServerDatabaseProvider` this is also causing the issue mentioned in #691. I believe it should have its own implementation but got stuck between a rock and a hard place.
This [SO anwser](https://stackoverflow.com/a/55811258) mentions three (actually only two if you take his performance comment into consideration) methods for paging in Oracle.

If you're happy with us only supporting Oracle 12c and above, the implementation is easy:
```sql
SELECT [Distinct ]FullName, Age 
FROM People
[ORDER BY FullName]
OFFSET 5 ROWS FETCH NEXT 14 ROWS ONLY
```
It just works, doesn't care if you omit `Order by` or include a `Distinct`.

If we need to support from 11g, I'm not sure how we'll know which version is used in order to cater for both.

**Comment by Ste1io** [comment]
> I've had huge success using Expando instead of ExpandoObject in one of my projects. See https://weblog.west-wind.com/posts/2012/feb/08/creating-a-dynamic-extensible-c-expando-object

Ironically, I've used the exact same concept in multiple projects myself also since reading Rick's article ~2018. 😆 I don't see that as being a viable solution in this case. either, though.

> In hindsight, only uppercasing would've been the issue then. In order to cater for all these flavors of database objects, it's gonna get messy very soon.

Agreed...realized this fairly quickly and didn't go down that route.

> If we need to support from 11g, I'm not sure how we'll know which version is used in order to cater for both.

[11g has been obsolete](https://support.oracle.com/knowledge/Oracle%20Cloud/2761482_1.html) for years, with [12c at EOL since 2022](https://support.oracle.com/knowledge/Oracle%20Database%20Products/742060_1.html) at this point. Even 12c support seems like an unnecessary burden if it brings with it any additional complexity to the library. Any thoughts here @pleb and @asherber?

> 1. Make the underlying dictionary, of the ExpandoObject, case insensitive.
> 2. Produce the correct casing on creation by convention.
> 3. Query the expected casing from the result based on db implementation flavor.

Ignoring case seems it could introduce some edge cases since it would no longer follow the same casing rules a strongly typed classs's properties follow. Most likely mapping the poco's property names (which would match the dynamic object keys) to the db column names by convention. The escaping methods would need to lookup the cached column mapping to determine if the string should be escaped, or not. Otherwise we're still stuck here, with one - but not both - being valid:

```cs
DB.Insert(_note);
var tblNote = DB.Provider.EscapeTableName(pd.TableInfo.TableName);
var entity = DB.Fetch<dynamic>($"SELECT * FROM {tblNote}").First();

entity.Text += " was updated"; // must be escaped PascalCase for class Note
DB.Update("Note", "Id", (object)entity); // must be UPPERCASE for default convention w/out quotes
```

Definitely needs some more thought on my end also. Will look at it some more this afternoon/evening once I'm home for the day.

**Comment by Curlack** [comment]
> Thinking to use string templating algorithm utility method in either new class or ParameterHelper utility.

Scratch that, I'm just gonna override the QueryTests methods.

**Comment by Ste1io** [comment]
> Scratch that, I'm just gonna override the QueryTests methods.

For the tests using a db-specific syntax, that's usually best. Some tests switch on the provider type inline if it's a minor change that doesn't necessitate overriding the method completely; either is fine.

**Comment by Curlack** [comment]
Quick question: Should output parameters also be listed in the `LastCommand` string?
I was able to include the RefCursor for the stored procs, but got "NullReferenceException" since output parameter doesn't have a value. So was thinking:
```c#
public string FormatCommand(IDbCommand cmd)
{
    //Exclude output parameters from the arguments list
    return FormatCommand(cmd.CommandText, (from IDataParameter parameter in cmd.Parameters where parameter.Direction.HasFlag(ParameterDirection.Input) select parameter.Value).ToArray());
}
```

**Comment by Curlack** [comment]
> Quick question: Should output parameters also be listed in the `LastCommand` string? I was able to include the RefCursor for the stored procs, but got "NullReferenceException" since output parameter doesn't have a value. So was thinking:
> 
> ```cs
> public string FormatCommand(IDbCommand cmd)
> {
>     //Exclude output parameters from the arguments list
>     return FormatCommand(cmd.CommandText, (from IDataParameter parameter in cmd.Parameters where parameter.Direction.HasFlag(ParameterDirection.Input) select parameter.Value).ToArray());
> }
> ```

Nevermind

**Comment by Curlack** [comment]
> Quick question: Should output parameters also be listed in the `LastCommand` string? I was able to include the RefCursor for the stored procs, but got "NullReferenceException" since output parameter doesn't have a value. So was thinking:
> 
> ```cs
> public string FormatCommand(IDbCommand cmd)
> {
>     //Exclude output parameters from the arguments list
>     return FormatCommand(cmd.CommandText, (from IDataParameter parameter in cmd.Parameters where parameter.Direction.HasFlag(ParameterDirection.Input) select parameter.Value).ToArray());
> }
> ```

It was the method below it causing the issue. It's when we call `args[i].GetType().Name` on a null. I'll just append "Unknown Type" to prevent it from falling over.

**Comment by Curlack** [comment]
How can I prevent the RefCursor parameter from being parameterized, i.e. "p_out_cursor" becomes ":p_out_cursor" via `AddParameter()` method?

**Comment by Curlack** [comment]
> How can I prevent the RefCursor parameter from being parameterized, i.e. "p_out_cursor" becomes ":p_out_cursor" via `AddParameter()` method?
> 
> I've tried reflection inside `OracleDatabaseProvider` during `PreExecute`, but that felt like a hack, so I scrapped that idea.
> 
> Now I'm trying something in `Database`, looks better, but not sure what the logic should be. I can't check for output parameter 'cause some out parameters might need the param prefix e.g. SqlServer I think. I can check the DbType (RefCursor is 121) or I can check the name. Here's what I have so far:
> 
> ```cs
> // Support passed in parameters
> if (value is IDbDataParameter idbParam)
> {
>     if (cmd.CommandType == CommandType.Text)
>         idbParam.ParameterName = cmd.Parameters.Count.EnsureParamPrefix(_paramPrefix);
>     else if (idbParam.ParameterName?.StartsWith(_paramPrefix) != true &&
>         //Don't touch output cursor parameters. This is specific to Oracle stored procedures
>         !string.Equals(idbParam.ParameterName, "p_out_cursor", StringComparison.OrdinalIgnoreCase))
>     {
>         idbParam.ParameterName = idbParam.ParameterName.EnsureParamPrefix(_paramPrefix);
>     }
> 
>     cmd.Parameters.Add(idbParam);
> }
> ```
> 
> Am I on the right track or is there a better way?

Looks like the overall rule should be to not touch output parameters. Please correct me if I'm wrong here.

**Comment by Curlack** [comment]
> Looks like the overall rule should be to not touch output parameters. Please correct me if I'm wrong here.

Ok scratch all of the above. Seems like it's Oracle that's full of it. It's correct that Oracle parameters are prefixed by a colon EXCEPT for stored procedures! I've been struggling with this one the whole day now....A day in the life of a developer hahaha

**Comment by Ste1io** [comment]
Just getting back to pc so catching up. Oracle package has an `OracleType` enum for reg params iirc. I was thinking add that as an overload.

**Comment by Curlack** [comment]
> Just getting back to pc so catching up. Oracle package has an `OracleType` enum for ref params iirc. I was thinking add that as an overload.

I've added the OracleDbType.RefCursor command type, if that's what you mean.

**Comment by Curlack** [comment]
FYI: Some goodies coming your way. See #707, hope it doesn't clash with some of the work you've been doing. Just glad I found the issue with the stored procedures.
Anyway I'm gonna call it a night.

**Comment by Ste1io** [comment]
Sounds great @Curlack. I'll look over things today.

**Comment by Ste1io** [comment]
@Curlack I may have figured out how to handle the escaping bit.. And it's actually fairly elegant if I may be so bold. Hopefully I'll have something later tonight for you to look at.

**Comment by Curlack** [comment]
In the beginning it sounded like a bad idea to ignore case, based on the reasons you mentioned [here](https://github.com/CollaboratingPlatypus/PetaPoco/issues/699#issuecomment-1774101484), but if Oracle is in case-insensitive mode, the code will just be modeled to the database and this will only be used for dynamic pocos. So I'm thinking we only need to cater for some one more dynamic type that allows you to specify whether to ignoreCase on creation and we should be done. This can even be the default dynamic type, but that would be a breaking change, just saying.

Consider the following `OracleExpandoObject` class:
```c#
public sealed class OracleExpandoObject : DynamicObject, IDictionary<string, object>
{
    private readonly IDictionary<string, object> _dictionary;
    public OracleExpandoObject() : this(false) { }
    public OracleExpandoObject(bool ignoreCase)
    {
        _dictionary = ignoreCase ?
            new Dictionary<string, object>(StringComparer.OrdinalIgnoreCase) :
            new Dictionary<string, object>();
    }

    #region IDictionary<string, object> Implementation
    public object this[string key] { get => _dictionary[key]; set => _dictionary[key] = value; }
    public ICollection<string> Keys => _dictionary.Keys;
    public ICollection<object> Values => _dictionary.Values;
    public int Count => _dictionary.Count;
    public bool IsReadOnly => _dictionary.IsReadOnly;
    public void Add(string key, object value) => _dictionary.Add(key, value);
    public void Add(KeyValuePair<string, object> item) => _dictionary.Add(item);
    public void Clear() => _dictionary.Clear();
    public bool Contains(KeyValuePair<string, object> item) => _dictionary.Contains(item);
    public bool ContainsKey(string key) => _dictionary.ContainsKey(key);
    public void CopyTo(KeyValuePair<string, object>[] array, int arrayIndex) => _dictionary.CopyTo(array, arrayIndex);
    public IEnumerator<KeyValuePair<string, object>> GetEnumerator() => _dictionary.GetEnumerator();
    public bool Remove(string key) => _dictionary.Remove(key);
    public bool Remove(KeyValuePair<string, object> item) => _dictionary.Remove(item);
    public bool TryGetValue(string key, [MaybeNullWhen(false)] out object value) => _dictionary.TryGetValue(key, out value);
    IEnumerator IEnumerable.GetEnumerator() => ((IEnumerable)_dictionary).GetEnumerator();
    #endregion

    public override bool TryGetMember(GetMemberBinder binder, out object result) =>
        _dictionary.TryGetValue(binder.Name, out result);
    public override bool TrySetMember(SetMemberBinder binder, object value)
    {
        _dictionary[binder.Name] = value;
        return true;
    }
}
```
It's not as badass as ExpandoObject, but support the ways we're using it e.g.
```c#
IDictionary<string, object> a = new OracleExpandoObject(true);

/*
a.Add("Test", "One");
a.Add("TEST", "Two"); //<- This will still error as per usual <DuplicateKey>
*/

a["Test"] = "One";
a["TEST"] = "Two";
Console.WriteLine(a["TesT"]);

dynamic b = a;
b.Test = "Three";
b.TEST = "Four";
Console.WriteLine(b.TesT);

/* Output:
 * Two
 * Four
 */
```

Off course it's not polished yet and we'll have to do load tests to see how it behaves under pressure. It shouldn't perform worse than `ExpandoObject` and should be thread safe (I'm seeing `LockObject` in MS code).

As for the Core, we're gonna need a provider flag to `IgnoreCase` for Oracle. We could maybe support some custom value in the `connectionstring`. This flag would then be passed in the constructor when creating an instance. There's also couple of places (4 it seems) where we check if the type is `ExpandoObject`, that must also take `OracleExpandoObject` into consideration.

So this is the extend of my research. Let me know what you guys think. Piece out.

**Comment by Curlack** [comment]
I just realized that whether the columns are case sensitive or not, it's not PP's problem. We're already covering all the basis and can handle both scenarios. It's up to the user to query the db poco, whether all caps or specific case, dynamic or strongly typed, correctly. All we need to do is fix how the tests are handling the db results and maybe add a case sensitive table for Oracle specific test and call it a day. Even escaping the names must remain noop to leave the user in full control (off course he can still override it if he wants to).

**Comment by Curlack** [comment]
> @Curlack I may have figured out how to handle the escaping bit.. And it's actually fairly elegant if I may say so myself. Hopefully I'll have something later tonight for you to look at.

I'm anxious to see what you came up with. Were you able to make progress with this? Because as you may have seen, I'm in two minds atm. 🤪

**Comment by Curlack** [comment]
If we let PP handle all of this, the only issue is a poco without explicit columns.

**Comment by Ste1io** [comment]
> > @Curlack I may have figured out how to handle the escaping bit.. And it's actually fairly elegant if I may say so myself. Hopefully I'll have something later tonight for you to look at.
> 
> I'm anxious to see what you came up with. Were you able to make progress with this? Because as you may have seen, I'm in two minds atm. 🤪

Have had little time this week thanks to a heavier-than-usual load from my usual work, @Curlack. I'll be free again this weekend. Sorry for the delay on my end. :(

**Comment by Curlack** [comment]
I'm seeing how far I can take the [ExpandoPoco](https://github.com/CollaboratingPlatypus/PetaPoco/issues/699#issuecomment-1778207981) idea mentioned earlier. Looks like I would need an `IgnoreCase` property in both the Provider (for `EscapeSqlIdentifier`) and Mapper (for `GetFactory`). Is there a way to sync these objects? For example `OraclDatabaseProvider` will override `IgnoreCase` (from `IProvider`), so when `Database` realizes there is no mapper, it defaults to `OracleConventionMapper` and set its `IgnoreCase` property to be the same as `OracleDatabaseProvider`'s. In this case the mapper would have liked `IgnoreCase` to be a constructor parameter, but has to remain an object with a parameterless constructor due to type constraints.

I managed to get 13 of the 15 failing tests to pass, but not yet happy with the changes, hence no PR yet.

**Comment by Ste1io** [comment]
Been a hellacious week with work; appreciate the patience. Will catch up with things here today.

**Comment by Curlack** [comment]
> Been a hellacious week with work; appreciate the patience. Will catch up with things here today.

No problem.

I've had some time on my hands and took these failing tests to town. I was able to make all of them pass, but had to dive deep into the heart of some core files. Changes are not too hectic though.

I decided to split the `OracleTestProvider` into 2 parts like what you did with the `SqlServerSystemDataTestProvider` and SqlServerMSDataTestProvider`. So we now have (well almost finished with that) `OracleOrdinaryTestProvider` and OracleDelimitedTestProvider`.

The gist of it is to ensure we cover the scenario where the user is having a case-sensitive database (delimited), which is also the way all the other databases work AND we cover the scenario where the user is having a case-insensitive database (ordinary). Technically the other databases also work like this except everything in Oracle is capitalized.

I'm hoping to give you a PR tonight still. I've gone outside the box with this one, so there's a chance that some changes are gonna raise eyebrows hahahaha. Hope at least it gives birth to some awesome ideas from thereon.

**Comment by Ste1io** [comment]
Sounds good, looking forward to seeing it.

I'm really thinking our best bet is to apply a global configuration option for Oracle for this, which is overridden on a per-property/column basis using an attribute/columninfo config. This would allow us to retain case sensitivity, and internally know whether or not to quote table names and sql identifiers. This would also allow all current functionality with dynamic types to work without forcing the typed poco to follow a convention.

**Comment by Curlack** [comment]
@Ste1io I ran all the integration tests just as a sanity check, expecting all to pass now, but no dice.

I'm getting the following error on a couple of the MSAccess tests:
```
Message: 
System.InvalidOperationException : Command parameter[2] '' data value could not be converted for reasons other than sign mismatch or data overflow.

---- System.Data.OleDb.OleDbException : Multiple-step OLE DB operation generated errors. Check each OLE DB status value, if available. No work was done.

  Stack Trace: 
OleDbCommand.ExecuteCommandTextErrorHandling(OleDbHResult hr)
OleDbCommand.ExecuteCommandTextForSingleResult(tagDBPARAMS dbParams, Object& executeResult)
OleDbCommand.ExecuteCommandText(Object& executeResult)
OleDbCommand.ExecuteReaderInternal(CommandBehavior behavior, String method)
OleDbCommand.ExecuteNonQuery()
<>c.<ExecuteNonQueryHelper>b__315_0(IDbCommand c) line 3226
Database.CommandHelper(IDbCommand cmd, Func`2 executionFunction) line 3249
Database.ExecuteNonQueryHelper(IDbCommand cmd) line 3226
Database.ExecuteInsert(String tableName, String primaryKeyName, Boolean autoIncrement, Object poco) line 2156
Database.Insert(Object poco) line 1929
DeleteTests.Delete_GivenPoco_ShouldDeletePoco() line 63
----- Inner Stack Trace -----
```

I'm fairly sure it's not related to my changes and Google is of no help.
It seems to be only the **deletes** and **inserts** causing this.
Have you had this issue before?

**Comment by Ste1io** [comment]
Yes...not sure what causes it, but restarting the msaccess service fixes it.

**Comment by Curlack** [comment]
I don't think you mean "windows service", but here it goes...
How do I do that?

**Comment by Curlack** [comment]
Btw, quickly trying to retype everything from my head. Wanted to undo a local commit and clicked "Reset (--hard)" by accident and like magic all my changes gone, ai! Still have some big parts of the code changes left to work with. Hold thumbs.

PS: I'm, still crying little bit.

**Comment by Ste1io** [comment]
> Btw, quickly trying to retype everything from my head. Wanted to undo a local commit and clicked "Reset (--hard)" by accident and like magic all my changes gone, ai! Still have some big parts of the code changes left to work with. Hold thumbs.
> 
> PS: I'm, still crying little bit.

OK, don't freak out... :) any other git commands besides that one? and any additional flags with the reset command?

**Comment by Ste1io** [comment]
Don't make any commits. If you have any changes on your working tree, stash them. Then run `git reflog`, and grab the commit for the missing commit. Then make a new branch (like "recovered-branch") using that commit hash with `git branch recovered-branch <hash>` Checkout that branch, and let me know if your changes are there (`git checkout recovered-branch`).

**Comment by Ste1io** [comment]
> I don't think you mean "windows service", but here it goes... How do I do that?

I was referring to restarting the msaccess db instance.

**Comment by Curlack** [comment]
> Don't make any commits. If you have any changes on your working tree, stash them. Then run `git reflog`, and grab the commit for the missing commit. You might need to guess at this one, but assuming you haven't done anything since the reset, it's probably the last commit hash). Then make a new branch (like "recovered-branch") using that commit hash with `git branch recovered-branch <hash>` Checkout that branch, and let me know if your changes are there (`git checkout recovered-branch`).

![image](https://github.com/CollaboratingPlatypus/PetaPoco/assets/298525/8c2081f0-df12-49cb-bae0-f85e3e5eff19)

Man, you know your stuff! Unfortunately I don't see any recovered changes, but also I lost changes I haven't committed yet (local only). Anyway almost done recalling from memory. Thanks bro!

**Comment by Curlack** [comment]
Ok so all tests now pass, see #709.
Hope the branch issues I had on my local branch/fork didn't bleed into the PR.

Calling it a night, tired of doing these changes twice hahaha.
Enjoy!

**Comment by Curlack** [comment]
Found weird issue when running tests, requiring custom provider and/or mapper in parallel with tests using the defaults. It seems the handling of the data, especially in my case (TitleCase vs UPPERCASE) bleeds into one another.
On further investigation, turns out that it's not right for PocoData to use a global cache (`_pocoDatas`), but not take the mapper into consideration. Seems like cache key should be `Type` AND `mapper` for 'PocoData`.

I bet the same thing is true for `PocoFactories` cache.

It can be reproduced fairly easily by debugging two tests in parallel, as long as they both use the same type poco e.g. Note and different mappers. Whoever gets to `ForType` first will dictate the way queries get generated.

Something like this will do:
```c#
private class CustomConventionMapper : ConventionMapper
{
    public CustomConventionMapper() => MapColumn = Customize(MapColumn);

    private Func<ColumnInfo, Type, PropertyInfo, bool> Customize(Func<ColumnInfo, Type, PropertyInfo, bool> original)
    {
        return (ci, t, pi) =>
        {
            if (!original(ci, t, pi)) return false;

            ci.ColumnName = ci.ColumnName.ToUpperInvariant();
            return true;
        };
    }
}
```

The issue came up in the `Fetch_ForDynamicTypeGivenSql_ShouldReturnValidDynamicTypeCollection` test (one of many). The line giving the error was in `QueryTests.AddOrders` when inserting a new `Person`.

I can either run all "Ordinary" or "Delimited" tests in parallel, but not all together...Pop goes the weasel!

I think my changes are ready to be committed. Can I create the PR or would you like me to take a stab at fixing this issue beforehand?

PS: I still having those pesky commits attached to my tail.

**Comment by Curlack** [comment]
I was able to flatten the history (sync fork/fetch/pull/rebase), unfortunately I wasn't able to get rid of the commits since 22 Oct.

This is what my branch look like atm.
![image](https://github.com/CollaboratingPlatypus/PetaPoco/assets/298525/6a8b4245-7b4f-43aa-9a82-f256784e0cc8)

Tried checkout and cherry picking again, but was losing too many changes when I looks at the diffs, so just rebased to origin.
Tried a couple of things, got scared and stopped. Kept me up till 3AM this morning.
So I'm still picking up these commits.

![image](https://github.com/CollaboratingPlatypus/PetaPoco/assets/298525/74a74e97-0c85-4194-81c2-7617715d84ab)
![image](https://github.com/CollaboratingPlatypus/PetaPoco/assets/298525/cb25755b-0932-47fc-9659-0cd42e1c0678)

Anything I should/can do to try and get rid of it (without losing changes)?

**Comment by asherber** [comment]
On the PocoData cache issue: https://github.com/CollaboratingPlatypus/PetaPoco/issues/517

**Comment by asherber** [comment]
On the git issue: Which branch in your repo, and what are you trying to do with it?

**Comment by Curlack** [comment]
My `feature/oracle-support` branch is apparently 13 commits behind when 10 of them have already been merged into the upstream branch of the same name. The original issue came in when I didn't sync my fork after those merges.
I want to know if there is something I can do to have my PR only contain the changes not yet merged.

In the meantime I've thinking about making a backup of the code on my drive, recreating the fork and copying the code over the download (should only see the differences). Would that work?

**Comment by asherber** [comment]
13 commits behind what? What branch do you want to merge into, and which 3 commits do you want to keep?

**Comment by asherber** [comment]
On git: The O'Reilly Git Pocket Guide is an excellent intro, even if it is 10 years old. And all of Git Pro is [free online](https://git-scm.com/book/en/v2)

Every git commit has a hash value (the 7 chars on the right of the GH screenshot above). Which three commits do you want to keep?

As for your comment about making a backup of the code on your drive, git calls that a clone, and it's the usual way of working with a repo. You forked PetaPoco into your account, now you clone it and work locally. All your dev work and commits are made locally, and then when you're ready they're pushed to your fork on GitHub. Then you open a pull request to get those changes merged to the main PetaPoco repo.

**Comment by Curlack** [comment]
Thanks so much. I'll study that in my spare time.
I wanted to keep bbcfcb0, 4c84dea and b0649a0, but I realize I can't do that otherwise I'll lose changes again. So what I'll do is start a new fork, clone. Copy my files over to the new folder from the backup and then go through the changes and do proper commits before the PR. Thanks again for the assist.

**Comment by asherber** [comment]
Hold on, there might be an easier way.

**Comment by asherber** [comment]
An important thing to keep in mind is that a git commit is not (just) a changeset. Technically, it's defined by the hash value. When you do things like rebasing and force pushing, you're creating entirely new commits which happen to have the same content as other commits. Now when you try to merge, git sees new commits (hashes it doesn't know about), but it will also understand what changes need to be made to the source files.

For example, I see your commit 31bcf, "Cater for optional parameter prefix". Looks like you rebased or did something else which created 0dfc8 -- same content, different commit. Depending on what you're merging, git will try to preserve both commits, but all that does is pollute the commit history (which we try to avoid) -- if they've got the same content, you're actually not introducing any other changes.

Take a look at https://github.com/CollaboratingPlatypus/PetaPoco/compare/feature/oracle-support...asherber:foo  Nevermind the list of commits -- does that represent the contents of what your want to merge?

**Comment by Curlack** [comment]
Eureka! That's exactly what I was after, unfortunately I saw your message too late. Started new fork already.
I couldn't unclutter the commit history. Did you do that straight from the web UI?

**Comment by asherber** [comment]
The GitHub UI is not good for any kind of branch management. What I did locally was create a new branch from 719731 and then cherry pick `bbcfc` and `b0649`.

Here are git commands you could use locally to accomplish this, using your existing `oracle-support` branch. Note that this will throw away any commits on your branch earlier than `71973`. 

```csharp

git checkout feature/oracle-support

git reset --hard 71973

git cherry-pick bbcfc

git cherry-pick b0649

git push --force
```

I do many git things from the command line, but a GUI tool is invaluable to see the branch relationships. My personal favorite is https://gitextensions.github.io/

**Comment by Curlack** [comment]
At long last success. All Oracle tests are passing (#710). @asherber I only needed the cache fix and a custom provider and mapper for the "Ordinary" identifiers case. As far as dynamic object casing is concerned, created overrides for the failing tests to use the correct casing coming from the poco factory, which the user would be aware of since he's the one that created the custom provider and mapper.
This time touch points in the core was kept to a minimum and all the "Delimited" tests were able to pass with out of the box functionality without a sweat.

Viva PetaPoco! 😄 ✌️

**Comment by Ste1io** [comment]
> Found weird issue when running tests, requiring custom provider and/or mapper in parallel with tests using the defaults. It seems the handling of the data, especially in my case (TitleCase vs UPPERCASE) bleeds into one another. On further investigation, turns out that it's not right for PocoData to use a global cache (`_pocoDatas`), but not take the mapper into consideration. Seems like cache key should be `Type` AND `mapper` for 'PocoData`.

@Curlack what's the status on this? If this is indeed a bug, could you file an issue for it specifically so it's not forgotten? That should be patched separately outside of the feature branch, though, as a bugfix. Still have a lot of catching up to do on here, so apologies in advance if I'm touching on something that's already been handled.

On the plus side I was able to spent most the morning pulling last week's thoughts and codes together so I have something coherent for you two to rip into. :)

**Comment by Curlack** [comment]
@Ste1io I'll submit another PR for this once the core change has been finalized on the develop branch. Issue has already been logged (See #517). You can follow our discussion in PR #711.

**Comment by Ste1io** [comment]
@Curlack, I've got an early day tomorrow and have to get off shortly, so I'll post the discussion I mentioned in my last message tomorrow. It details quite a few things that I encountered over the weekend while exploring some approaches to the quoting issue with runtime objects like anonymous types and dynamic objects.

Regarding my solution, however, here's a *very* abbreviated synopsis, so you and @asherber can mule over it between now and then.

Several months ago, I created a helper class, [**`PocoDataHelper<T>`**](https://github.com/Ste1io/PetaPoco/blob/feature/PocoDataHelper/PetaPoco/Core/PocoDataHelper.cs) designed to act as a bridge of sorts (I guess decorator would be the more correct term) between `PocoData` and the escape methods in `IProvider`. Goals:

- Remove the boilerplate code currently necessary (creating a `PocoData` instance, followed by verbose calls to the provider, which wrap even more verbose linq queries for looking up the column name)
- Accommodate more readable and expressive code in my personal projects
- Promote more refactor-friendly and maintainable code, leveraging expressions to my advantage to eliminate the use of strings wherever possible
- Provide intellisense support when referencing typed columns (which comes for free thanks to expressions)
- Make it easier for the end user - aka myself - to do the right thing because it's easier than cutting corners or c/ping from another section of code 
- Do all this in a non-intrusive manner by applying the open/closed principle

Expressions do not incur the typical performance hit as there is no compilation being done; I'm simply extracting the property name from the expression, and doing a lookup for it in the `PocoData.Columns` cache. An optional `IMapper` can optionally be injected during construction with the same default that the core library uses already. I also am testing a version today that sets the `PocoData.TableData.TableName` property when an instance is constructed from the non-generic base `PocoDataHelper` class constructor, to further eliminate repeated code when used with runtime objects.

Escaping/quoting behavior is configured on a per-instance basis for flexibility, with the ability to override it as needed when referencing an identifier, such [as in this test](https://github.com/Ste1io/PetaPoco/blob/feature/PocoDataHelper/PetaPoco.Tests.Integration/Databases/QueryTests.cs#L232-L234):

```cs
var pd = new PocoDataHelper<Order>(DB, escapeIdentifiers: false);
var sql = $"SELECT * FROM {pd.GetTableName(escapeIdentifier: true)} " +
          $"WHERE {pd.GetColumnName(c => c.Status, escapeIdentifier: true)} = @0";
```

My intentions were to finish full test coverage and do some perf profiling before bringing it to the table as something to consider adding, but as it currently offers the most feasible and safest solution to handling the intricacies with Oracle, I'm letting her enter the fray early. I've also been using it in two projects since originally writing her, with no problems.

I'm sure you have an idea the direction I'm suggesting by now, but just to be clear, a few considerations below. The current implementation I'm using for identifier escaping could be applied to the Oracle-specific configurations also, with sane defaults. Currently I'm leaning towards something along the lines of the following:

**QuotingPolicy:** Never (default), Always, WhenNecessary
**CasingPolicy:** FoldToUpper (default), FoldToLower, PreserveCase

This could be set on a global level,

```cs
var db = DatabaseConfiguration.Build()
                              .UsingConnectionStringName("Oracle")
                              .UsingQuotingPolicy(QuotingPolicy.Never)     // Never (default), Always, WhenNecessary
                              .UsingCasingPolicy(CasingPolicy.FoldToUpper) // FoldToUpper (default), FoldToLower, PreserveCase
                              .Create();
```

which the Oracle-specific helper class would respect unless overridden for specific instances (per-table override) or call sites (per-column override).

Keep in mind, one reason there are currently so many overloads for the base CRUD functionality, is to pass through the needed data for that particular database operation, when dealing with a runtime type. An anonymous or dynamic type can't have attributes, obviously. So the primarykey gets passed in through the call to `Fetch`, `Query`, `Update`, and all the others.

Currently there's one Oracle-specific attribute: `SequenceName`, along with several others that haven't been brought up yet in these discussions, but do exist and will spring their ugly head the minute we think we got it. If I'm not using a typed class and attributes, how do we pass it through?

Option 1: More overloads, add a parameter for it, handle it at the core level
Option 2: Add it to our decorator class which extends `PocoDataHelper<T>`, perform any logic necessary there, then allow the user's method call to work with our fixed-up final value no different than any other provider.

This also has the added benefit of being able to specify something, such as quoted/unquoted, in a flexible manner even inline, with no adverse effects. True, it can be done manually, though I think we all agree that argument has nothing going for it given how error-prone it is (besides locking you into that dbms without a major overhaul that touches everything everywhere), As a trivial example just to illustrate:

```cs
// Quoted or unquoted column name, with following signature:
db.SomeOperation<T>(string tableName, string primaryKeyColumnName);

// At callsite:
dynamic myFooTable = //...
var pdh = new PocoDataHelper(db, myFooTable, "Id");
db.SomeOperation<dynamic>(tableName: "FooTable", pdh.GetColumnName("BarColumn", true)); // db: 'BarColumn'
db.SomeOperation<dynamic>(tableName: "FooTable", pdh.GetColumnName("BazColumn")); // db: 'BAZCOLUMN'
```

In closing:

- Yes, I'm introducing a new design pattern, which I'm usually hesitant to do. However, it improves the usability of the entire codebase, not just Oracle, and provides an extensible and flexible tool to accommodate Oracle's edge cases without introducing potential regressions from invasive core library changes.
- Implementing Oracle-specific behavior *from the outside*, without coupling it to core code that touches other providers, keeps ensures that none of the current providers are affected. This means no performance hit on other providers, and no change to the current API surface (out of necessity to deal with Oracle).
- Quoting logic would be handled here in `PocoDataHelper<T>` (or a derived `OraclePocoDataHelper<T>`), with the escaping method in `OracleDatabaseProvider` being a non-transformative pass-through method.

This was far less "brief" than I meant when I started, but hopefully it gives everyone something to chew on, and I'm looking forward to your thoughts.

I updated all synchronous test methods for both [PetaPoco.Tests.Integration/Databases/QueryTests.cs](https://github.com/CollaboratingPlatypus/PetaPoco/compare/development...Ste1io:PetaPoco:feature/PocoDataHelper#diff-52de05a8238b11d5a9026ea252dbd084efa7cc72e45d2974634353614f3f475f) and [SQLite's derived QueryTests](https://github.com/CollaboratingPlatypus/PetaPoco/compare/development...Ste1io:PetaPoco:feature/PocoDataHelper#diff-3dbe6ae2d18c62a24083eaf61e5f8c50a0e040402b5636cc3adf22341d1c03fd), with all tests passing. Please take a few minutes to look over the diffs for those two linked files, as well as the class itself (under 85 lines), in the context of Oracle's identifier handling especially with anonymous and dynamic types.

Commit history including diffs from test cases here - [click here](https://github.com/CollaboratingPlatypus/PetaPoco/compare/development...Ste1io:PetaPoco:feature/PocoDataHelper).

**Comment by Curlack** [comment]
I like this a lot. Can't wait to see the full version.

**Comment by Curlack** [comment]
Would this PocoDataHelper also be able to make edge cases like #657 easier to deal with?

**Comment by Ste1io** [comment]
I've only briefly skinned over the thread, and that was a few months ago, so can't say for sure until I can review it a little more thoroughly; at this point you probably would know better than I.

The property-to-column map is just a simple string lookup that saves PetaPoco's cached column on first access; no need for more than that in this case really. As with the column cache tho, it really depends on if the performance gains from not making repeated lookups to the global cache justified what (relatively little) extra data storage it incurred, as well as synchronizing cache flushes in case there's any long-living instances being used. I didn't want to invest too much time into caching until I had a chance to profile it though; what I originally wrote it for, it does damn well, so until now the rest was just a bonus.

Since it is self contained and only represents a single table per instance, I guess it could potentially help from what little I read, possibly with a couple minor adjustments. Just keep in mind it's still using the global cache as the source of truth on the first column lookup, if that makes any difference.
